Task №13.  (this task pushed not in correct branch)

jacsmedvedev opened the pull request:
First commit. I have not fulfilled the last condition (Предусмотрите
ситуацию, когда у одного автовладельца может быть несколько автомобилей)

jacsmedvedev committed:
Task №13.

First commit. I have not fulfilled the last condition (Предусмотрите
ситуацию, когда у одного автовладельца может быть несколько автомобилей)

jacsmedvedev committed:
Task №13 commit 2.

Добавил функцию добавления номеров машин у одного владельца.

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+ * */
+public class AutoOwnerDirectory {
+
+	Map<String,List<String>> _manAndCarNumber = new HashMap<String,List<String>>();
```
Почему без спецификатора доступа? Будет использоваться внутри пакета?

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	public Map<String, List<String>> getManAndCarNumber() {
+		return _manAndCarNumber;
```
Нужно возвращать неизменяемую мапу

strazhkoe replied:
Забыл за этот комментарий?

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+		return _manAndCarNumber;
+	}
+
+	public void setManAndCarNumber(Map<String, List<String>> manAndCarNumber) {
```
Думаю, что этот метод вообще стоит удалить

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+		_manAndCarNumber = manAndCarNumber;
+	}
+
+	public void putElement(String name, List<String> number){
```
плохое название метода, ни о чем не говорит. Тем более нет джавадока, описывающего аргументы

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+	}
+	
+	public void removeElement(String name){
+		_manAndCarNumber.remove(name);
```
Как узнать, успешно удалился или нет или его вообще там небыло?

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+	}
+	
+	public void editElement(String name, String newName, List<String> newNumber){
+		if(this._manAndCarNumber.containsKey(name) == true ){
```
== true лишнее

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+	public void editElement(String name, String newName, List<String> newNumber){
+		if(this._manAndCarNumber.containsKey(name) == true ){
+			this._manAndCarNumber.remove(name);
+			this._manAndCarNumber.put(newName,newNumber);
```
метод put перезатирает старое значение. Нет необходимости предварительно удалять. См. https://docs.oracle.com/javase/7/docs/api/java/util/Map.html#put(K,%20V)

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+		for (Entry<String, List<String>> entry: this._manAndCarNumber.entrySet()) {
+			if (entry.getValue().contains(number) == true){
+				String result =  entry.getKey();	
+				System.out.println(result + " is owner of tne car number " + number);
```
Результат нужно вернуть, а не печатать. Вдруг он не для печать нужен?

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,63 @@
+		}
+	}
+	
+	public void print() {
```
нужно перенести из этого класса

strazhkoe commented on src/task_from_teacher_13/CarNumberComparator.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.List;
+import java.util.Map;
+
+public class CarNumberComparator implements Comparator<Map.Entry<String, List<String>>> {
```
Джавадок!!!

jacsmedvedev committed:
Task №13 commit 3.

1.added modifier private to _manAndCarNumber field
2.deleted setManAndCarNumber method
3.rename method putElement to addNewElement
4.added notifications to the method removeElement
5.in method findName changed void to String
6.method print() removed to class Main
7.Added javadocs

Akros2016 commented:
What's wrong ?

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -23,41 +20,59 @@ public AutoOwnerDirectory() {
+	 * @param name car owner name 
+	 * @param number car number 
+	 */
+	public void addNewElement(String name, List<String> number){
```
Не намного лучше старого названия. Назови его например addCarOwner или что-то около этого

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -23,41 +20,59 @@ public AutoOwnerDirectory() {
+				System.out.println("Запись не удалена.Повторите попытку.");
+			}
+		} else {
+			System.out.println("Записи с именем " + name + " не найдено. Уточните данные.");
```
этот метод должен возвращать результат выполнения операции, а выводы в интерфейс должны быть сделаны в классе, отвечающим за общение с пользователем

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -23,41 +20,59 @@ public AutoOwnerDirectory() {
+	 * @param newName new car owner name
+	 * @param newNumber car number of new car owner
+	 */
 	public void editElement(String name, String newName, List<String> newNumber){
```
нет обратной связи, метод ничего не возвращает. А если такого элемента нет, то метод просто промолчит?

jacsmedvedev committed:
Task №13 commit 4.

- rename method addNewElement to addNewCarOwner;
- in methods addNewCarOwner(), removeElement(), editElement() changed
return type to boolean;
- added lambda expressions.

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,89 @@
+	 */
+	public boolean addNewCarOwner(String name, List<String> number) {
+		_manAndCarNumber.put(name, number);
+		if (_manAndCarNumber.containsKey(name)) {
```
Ты же только что сюда положил пару с этим ключем? Зачем теперь это проверять? Есть ли такие случаи, когда можно выполнить put, и пара не добавится в мапу?

strazhkoe replied:
весь этот if-else можно заменить на return _manAndCarNumber.containsKey(name). Но нужно ли это? См. комментарий выше

jacsmedvedev replied:
Ты же сам где то писал, что нужно проверять, успешно ли прошла операция и как-то об этом сигнализировать !

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,89 @@
+		} else {
+			return false;
+		}
+	}
```
Весь этот метод заменяется одной строчкой:
`
return _manAndCarNumber.remove(name) != null
`
Читай javadoc к интерфейсу Map, особенно те методы, что собираешься использовать

strazhkoe commented on src/task_from_teacher_13/AutoOwnerDirectory.java:
```diff
@@ -0,0 +1,89 @@
+		} else {
+			return false;
+		}
+	}
```
А удалить запись под ключем name? Ты создал новую, но не удалил старую

strazhkoe commented on src/task_from_teacher_13/CarNumberComparator.java:
```diff
@@ -0,0 +1,20 @@
+	
+	@Override
+	public int compare(Map.Entry<String, List<String>> o1, Map.Entry<String, List<String>> o2) {	
+		return o1.getValue().toString().compareTo(o2.getValue().toString());
```
а если в двух списках одни и те же номера будут идти в разном порядке эти списки будут различны? Это правильно?

jacsmedvedev replied:
То есть сначала отсортировать номера внутри самого листа  ?

strazhkoe replied:
Класс, как я смотрю, вообще не используется. Можно его удалить

jacsmedvedev committed:
Task №13 commit 5

- method getManAndCarNumber() returns unmodifiable map
- removeElement() - reduced the amount of code inside method
- editElement() - changed code inside method

TODO sort car numbers in the Lists

jacsmedvedev committed:
Task №13 commit 6

- added sorting car numbers in the List (for print)

jacsmedvedev committed:
Task №13 commit 6

- added sorting car numbers in the List (for print)
